pkg/serviceaccounts/controllers/docker_registry_service.go

```diff
@@ -242,16 +285,37 @@ func (e *DockerRegistryServiceController) getDockerRegistryLocations() []string
 		return []string{}
 	}
 	service := obj.(*kapi.Service)
+	result := []string{}
 
 	hasClusterIP := (len(service.Spec.ClusterIP) > 0) && (net.ParseIP(service.Spec.ClusterIP) != nil)
 	if hasClusterIP && len(service.Spec.Ports) > 0 {
-		return []string{
+		result = append(result, []string{
 			net.JoinHostPort(service.Spec.ClusterIP, fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
 			net.JoinHostPort(fmt.Sprintf("%s.%s.svc", service.Name, service.Namespace), fmt.Sprintf("%d", service.Spec.Ports[0].Port)),
+		}...)
+	}
+
+	// Generate dockercfg entry for value of OPENSHIFT_DEFAULT_REGISTRY (if set)
```

Comment:
@miminar @deads2k this makes is a little bit more robust and deal with OPENSHIFT_DEFAULTS_REGISTRY variable (test updated as well).

The supported mechanisms are basically:

1) You expose the docker-registry via Route (`registry.foo.local`) -> we generate entry for `registry.foo.local` (with no ports, assuming docker will figure out whether it is 80 or 443).
2) You expose via Route but create Routers that routes into a non-standard port (eg. `registry.foo.local:5000`). In that case, you have to set a `OPENSHIFT_DEFAULT_REGISTRY=registry.foo.local:5000) as we can't tell what port the router is using.
3) You don't expose any route, just set `OPENSHIFT_DEFAULT_REGISTRY` - in that case we just create secret for the URL you provided (and regenerate on every master restart).

The 1) and 2) regenerate secrets everytime you change the Route hostname on live server. The 2) also allows to do crazy thing like set the `OPENSHIFT_DEFAULT_REGISTRY=docker-registry.default.svc.cluster.local` but also expose it via public route. Both will get the secrets generated properly.